code/modules/food_and_drink/cakes.dm

```diff
@@ -24,62 +24,482 @@
+		var/topping //the topping the player is adding (stored as a string reference to an icon_state)
+		var/pendinglight //a variable referenced later to check if the light source on a cake needs to be updated
+		if(istool(W, TOOL_CUTTING | TOOL_SAWING))
+			slice_cake(W,user)
+			return
+		else if(istype(W,/obj/item/reagent_containers/food/drinks/drinkingglass/icing))
+			frost_cake(W,user)
+			return
+		else if(istype(W,/obj/item/reagent_containers/food/snacks/cake/custom))
+			stack_cake(W,user)
+			return
+		else if(cake_candle && !(litfam) && (W.firesource))
+			src.ignite()
+			W.firesource_interact()
+			return
+		else
+			var/list/returns = check_for_topping(W) //if the item used on the cake wasn't handled previously, check for valid toppings next
+			if(returns[1] == 0) //if the item wasn't a valid topping, perfom the default action
+				..()
 				return
-			if (!icing)
-				boutput(user, "<span class='alert'>You need to add icing first!</span>")
+			topping = returns[1]
+			pendinglight = returns[2]
+
+			//adding topping overlays to the cake. Yay :D
+			if(src.sliced) //if you add a topping to a sliced cake, it updates the icon_state to the sliced version.
+				topping = replacetext("[topping]","cake[clayer]","slice")
+			if(topping && !(src.GetOverlayImage("[topping]"))) //actually adding the topping overlay to the cake
+				var/image/toppingoverlay = new /image('icons/obj/foodNdrink/food_dessert.dmi',"[topping]")
+				toppingoverlay.alpha = 255
+				src.UpdateOverlays(toppingoverlay,"[topping]")
+				user.u_equip(W)
+				qdel(W)
+				if(pendinglight)
+					src.ignite()
 				return
-			boutput(user, "<span class='notice'>You cut the cake into slices.</span>")
-			var/makeslices = src.amount
-			while (makeslices > 0)
-				var/obj/item/reagent_containers/food/snacks/cake/custom/P = new src.type(get_turf(src))
-				P.food_effects += src.food_effects
-				P.sliced = 1
-				P.amount = 1
-				P.w_class = 1
-				P.icon_state = "cake_slice"
-				P.quality = src.quality
-				P.name = src.name
-				P.desc = src.desc
-				P.food_color = src.food_color
-				P.update_icon(1)
-				P.icing_color = src.icing_color
-				P.update_icing(1)
-				src.reagents.trans_to(P, src.reagents.total_volume/makeslices)
-				P.pixel_x = rand(-6, 6)
-				P.pixel_y = rand(-6, 6)
-				makeslices--
-			qdel (src)
-		if(istype(W, /obj/item/reagent_containers/glass/bottle/icing) && !(icing) &&(!sliced) && (W.reagents.total_volume))
-			if(W.reagents.total_volume == 50)
-				boutput(user, "<span class='notice'>You add the icing to the cake.</span>")
-				icing = 1
-				src.desc += "<br>It has " + W.reagents.get_master_reagent_name() + " icing."
-				icing_color = W.reagents.get_master_color()
-				src.update_icing(0)
-				W.reagents.trans_to(src, W.reagents.total_volume)
+
+
+	attack_hand(mob/user as mob)
+		if((user.a_intent == INTENT_GRAB) && (src.clayer >1)) //removing layers from cakes.
```

Comment:
I think we generally don't like intent based interaction with items but I'm also not sure *what* we prefer.

Reply:
multicontext, if the effort can be put in

Reply:
it required a weird workaround with the existing framework for multicontext menus, but i got it working!